static/app/views/organizationContext.tsx

```diff
@@ -1,5 +1,147 @@
+import {metric} from 'sentry/utils/analytics';
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import {useParams} from 'sentry/utils/useParams';
+import {useRoutes} from 'sentry/utils/useRoutes';
 
 export const OrganizationContext = createContext<Organization | null>(null);
+
+interface Props {
+  children: React.ReactNode;
+}
+
+/**
+ * There are still a number of places where we consume the lgacy organization
+ * context. So for now we still need a component that provides this.
+ */
+class LegacyOrganizationContextProvider extends Component<{value: Organization | null}> {
+  static childContextTypes = {
+    organization: SentryTypes.Organization,
+  };
+
+  getChildContext() {
+    return {organization: this.props.value};
+  }
+
+  render() {
+    return this.props.children;
+  }
+}
+/**
+ * Context provider responsible for loading the organization into the
+ * OrganizationStore if it is not already present.
+ */
+export function OrganizationContextProvider({children}: Props) {
+  const api = useApi();
+  const configStore = useLegacyStore(ConfigStore);
+
+  const {organizations} = useLegacyStore(OrganizationsStore);
+  const {organization, error} = useLegacyStore(OrganizationStore);
+
+  const hasMadeFirstFetch = useRef(false);
+
+  const lastOrganization: string | null =
+    configStore.lastOrganization ?? organizations[0]?.slug ?? null;
+
+  const routes = useRoutes();
+  const params = useParams<{orgId?: string}>();
+
+  // Determine orgSlug from the parameters
+  const orgSlug = params.orgId || lastOrganization;
+
+  const handleLoad = useCallback(() => {
+    if (!orgSlug) {
+      return;
+    }
+
+    metric.mark({name: 'organization-details-fetch-start'});
+    fetchOrganizationDetails(api, orgSlug, false, hasMadeFirstFetch.current);
+    hasMadeFirstFetch.current = true;
+  }, [api, orgSlug]);
+
+  // If the organization slug differs from what we have in the organization
+  // store reload the store
+  useEffect(() => {
+    // Nothing to do if we already have the organization loaded
+    if (organization && organization.slug === orgSlug) {
+      return;
+    }
+
+    handleLoad();
+  }, [orgSlug, organization, handleLoad]);
+
+  // Take a measurement for when organization details are done loading and the
+  // new state is applied
+  useEffect(
+    () => {
+      if (organization === null) {
+        return;
+      }
+
+      metric.measure({
+        name: 'app.component.perf',
+        start: 'organization-details-fetch-start',
+        data: {
+          name: 'org-details',
+          route: getRouteStringFromRoutes(routes),
+          organization_id: parseInt(organization.id, 10),
+        },
+      });
+    },
+    // Ignore the `routes` dependency for the metrics measurement
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [organization]
+  );
+
+  // Configure sentry SDK scope to have organization tag
+  useEffect(() => {
```

Comment:
Since this runs in a useEffect, it might happen that the children components triggering sentry related events won't have these properties applied as the effect will run after the component (for example if a child throws a sync error crashing the page, this effect will never run)

A solution would be to inline it like
```js
const initializedSentryRef = useRef(false)

if(!initializedSentryRef.current && organization){
  Sentry.configureScope...
  initializedSentryRef.current = true
}
```

Reply:
Oof yeah this is tricky you're right. This probably should just be built into the `fetchOrganizationDetails`. I'll move it there

Reply:
Okay I moved this out of here.